stablehlo/reference/Ops.cpp

```diff
@@ -157,6 +157,25 @@ Tensor evalOrOp(const Tensor &lhs, const Tensor &rhs, Type resultType) {
   return result;
 }
 
+Tensor evalPadOp(const Tensor &operand, const Tensor &paddingValue,
+                 ArrayRef<int64_t> edgePaddingLow,
+                 ArrayRef<int64_t> interiorPadding, Type resultType) {
+  Tensor result(resultType);
+  for (auto resItr = result.index_begin(); resItr != result.index_end();
+       ++resItr)
+    result.set(*resItr, paddingValue.get({}));
+  for (auto opItr = operand.index_begin(); opItr != operand.index_end();
+       ++opItr) {
+    SmallVector<int64_t> adjustedIdx(result.getType().getRank());
+    for (int64_t i = 0; i < operand.getType().getRank(); ++i)
+      adjustedIdx[i] =
+          edgePaddingLow[i] + (*opItr)[i] * (interiorPadding[i] + 1);
+    if (failed(verifyIndex(result.getType().getShape(), adjustedIdx))) continue;
+    result.set(adjustedIdx, operand.get(*opItr));
```

Comment:
1. No initialization of result Tensor. Ideally a `Tensor` has to be immutatble (https://github.com/openxla/stablehlo/issues/233).
2. IMO this looks more closer to the spec. 

```
#include<numeric>
 // not initialization of result
 SmallVector<int64_t> dims(operand.getType().getRank());
  std::iota(dims.begin(), dims.end(), 0);
  if (llvm::any_of(dims, [](int64_t dim)
        return (*resItr)[dim] !=
               edgePaddingLow[dim] + (*opItr)[dim] * (interiorPadding[i] + 1);
      }))
    result.set(*resItr, paddingValue.get({}));
  else 
    result.set(*resItr, operand.get(*opItr));
```

Please let me know if you have other concerns

Reply:
Though the current code is initializing the padding values first, I think it still stays close to the spec. I've considered writing without initializing the tensor, but the way the spec is written, though concise, needs to loop through both resultIdx and operandIdx resulting in quadratic complexity. If we initialize the result early, we run at O(max(operand, result)) linear time. I've also looked into your suggestion, but it has the problem of which tensor to iterate to make it linear: result or operand or both. I haven't considered how to do this in linear time in depth, so I may be missing something, so let me know if you have suggestions in mind.

Reply:
I agree with your observation. Checking if a result index matches with an expression of operand index and padding attributes requires iterating over all the operands. That would unnecessarily complicates the code.  Computing the result index from an operand index (which is what you are doing) seems right and concise. Please drop my suggestion.